compile.js

```diff
@@ -415,3 +481,60 @@ function getDefaultWriteTest(ctx, field) {
 function isPacked(field) {
     return field.options.packed === 'true';
 }
+
+function getTsType(field) {
+    let type = field.type;
+
+    if (fieldShouldUseStringAsNumber(field)) type = 'string';
+    else if (fieldTypeIsNumber(field)) type = 'number';
+    else if (field.type === 'bytes') type = 'Uint8Array';
+    else if (field.type === 'bool') type = 'boolean';
+
+    return field.repeated ? `${type}[]` : type;
+}
+
+function getMapTsType(fields) {
+    const key = getTsType(fields[0]);
+    const value = getTsType(fields[1]);
+    return {key, value};
+}
+
+function typeDef(type, name, fields = []) {
+    const unionProperties = {};
+    const properties = fields.map((field) => {
+        if (field.oneof) {
+            unionProperties[field.oneof] = unionProperties[field.oneof] || [];
+            unionProperties[field.oneof].push(field.name);
+        }
+
+        const type = getTsType(field);
+        const isRequired = field.required || field.repeated || field.map;
```

Comment:
```suggestion
        const isRequired = field.required;
```

`repeated` fields (and `map`) are inheritely optional as they can be provided with zero values.
Should we not reflect that in the JSDoc and makes all repeated and map fields optional ?